#113 time & lighting system

FAStre opened the pull request:
Build has been submitted and awaiting testing team approval. 

Closes #113 
Closes #526

FAStre committed:
Simple Day and Night Cycle System asset added

FAStre committed:
Merge pull request #278 from AoTTG-2/development

Update #113 with development

FAStre committed:
Added DayNightSystem object to all Scenes

FAStre committed:
Added DayNightSystem object

forgot to in previous commit

TheBoredProgrammer64 committed:
Optimized the asset

FAStre committed:
Added back deleted code and reinstated prefab

TheBoredProgrammer64 committed:
Added a field to TimeSwitcher.cs

TheBoredProgrammer64 committed:
Added a field to TimeSwitcher.cs

FAStre committed:
Update aottg2

FAStre committed:
resolved merge conflict

TheBoredProgrammer64 committed:
Update TimeSwitcher.cs

TheBoredProgrammer64 committed:
TimeSwitcher

TheBoredProgrammer64 committed:
TimeSwitcher script fixed

TheBoredProgrammer64 committed:
Made the setting visible and did few changes.

The slider doesn't work, for some reaason the game crashes instantly when you click on server settings.

TheBoredProgrammer64 committed:
Made changes.

FAStre committed:
Fixed UpdateTime() call from server settings page and added slider value debug message

-The slider as is is jut returning the value 0 from what ive een in debug messages.

-I think that some slight redesign, maybe moving UpdateTime to DayAndNightControl.cs might be better, will test soon.

FAStre committed:
Change current time slider now working

Added some comments to allow easy addition of more sliders and text inputs.

FAStre committed:
update aottg2.unity

FAStre committed:
Added exact time text input, temporarily disabled timeslider

-Very basic text input that allows users to enter an exact time, ie 17:34 and have the cycle move to that point
**a few issues have been noticed, to fix in next commits:**
1. When entering the server settings, the slider is updated every frame, and therefore overrides the text field, easy fix though.
2. Day seems too dark and night time to bright, Ive noticed the asset colours seem to have been reset, will look at past asset properties with hard git reset on local machine.
3.For some reason the moon is a purple box, will compare to previous versions with hard git reset locally, as the moon was working at some point, but some recent commits turned it into a pink/purple box.

FAStre committed:
Slider and text input now works together

FAStre committed:
Added Day length input field

FAStre committed:
update with development

FAStre committed:
reverted incorrect development merge

FAStre committed:
Merge branch 'development' into #113-Time-&-Lighting-System

FAStre committed:
Added button that activates daynight system

-At the moment only instantiates the prefab, still need to make it a toggle button

FAStre committed:
Reverted changes to PhotonServerSettings

For some reason they changed locally in a previous commit and i forgot to exclude them, this si changing it back to before i changed this file.

FAStre committed:
Merge branch 'development' into #113-Time-&-Lighting-System

FAStre committed:
Updated UI

-Toggle button added, InstantiateDayNightController.cs changed to work with a toggle
-Removed unneeded scene change and day start code from DayAndNightControl.cs

FAStre committed:
Fixed moon and lighting issue

- readded some old versions of assets, from when moon and lighting as still normal.
-Will remove useless addons near the end of issue completion

FAStre committed:
Added DayLength Control and removed unneeded code

FAStre committed:
Integrated default light as a prefab

-Now the default directional light is turned off when the dynamic time and lighting system is toggled on, to prevent lighting conflict.

FAStre committed:
Implemented time freeze button

-now the daynightcycle can freeze
-adjusted lightset prefab higher to match the default lighting

FAStre committed:
Skyboxes now change according to time of day

-3 phases, morning, midday and night have defined skyboxes that activate during that time.
-added a large amount of extra skyboxes, for use if more phases such as early morning, early afternoon, midnight etc phases of a day will be added.

FAStre committed:
Removed stardome and fixed reset to default lighting

-Stardome was of negligible use, and interfered with the aesthetic of night skyboxs, therefore removed from DayNightCycle Prefab
-Now when deactivating the dynamic daynight system, the default scene skybox is reinstantiated and all lighting settings reverted to scene defaults
-Instead of destroying and recreating light prefabs, now they are rather disabled and reenabled
-An error persists when changing scenes, need to reload variables on scene change.

FAStre committed:
UI labels updated and Utgard lighting fixed

-Text labels added to settings for time and lighting system
-Utgard lighting settings changed from skybox to colour
-edited times for when dawn, day and night appear

FAStre committed:
Added Moonlight and Updated night skybox

FAStre committed:
Edited floor names to allow daynight cycle automatic scaling and fitting

FAStre committed:
revert accidental camera prefab edit

FAStre committed:
Refactored code and fixed outside the walls error

-Now scene terrain renamed to "GroundTerrain" as jsut "Terrain" resulted in conflicts with dynamic time and lighting system instantiating in outside the walls map.
-Future maps should name their ground terrain/objects as 'GroundTerrain'

FAStre committed:
Added MoonCamera and Moon Layer

- Allows rendering of moon even on larger scenes, without increasing default render distance

FAStre committed:
Removed unnused skybox assets

FAStre committed:
Changed scenes ground names

FAStre committed:
Modified some time variables

FAStre committed:
Resolved Merge Conflicts

FAStre committed:
Fixed MoonCamera issue and added lighting check

FAStre committed:
Revert "Fixed MoonCamera issue and added lighting check"

This reverts commit b2d5c4ca1284d2a39a9976d6ade09b0e8311058b.

FAStre committed:
Update Day and Night Controller.prefab

FAStre committed:
Sonarcloud fixes

FAStre committed:
Fixed day shadows and added smooth colour transition

-Directional light colour changed to allow for sharper shadows at day/dawn
-Lerp added for smoother colour transitions

FAStre committed:
Fixed moon rendering incorrectly

-MoonCamera and MainCamera now have the same field of view, eliminating moon rendering issues while the player moved/looked at the moon from different angles

FAStre committed:
Added multiplayer compatibility

-Gameobject is now instantiated via PhotonNetwork.instantiate
-Prefab moved to Resources Folder to allow for PhotonNetwork.instantiate
-GameSettings.cs extended to include timesettings
-TimeSettings.cs added

FAStre committed:
disable default lighting in multiplayer instead of destroying it

FAStre committed:
PhotonNetwork instantiation now instantiates correctly

FAStre committed:
Added RPC calls for synching time and lighting in multiplayer

-Removed unneeded gameobject.find calls to optimise methods
-Edited Scenes to have default skyboxes set.
-SyncTime() added as a PUN RPC

FAStre committed:
Removed unnecessary RPC's and restructured project files

-Added an extra transitionary skybox for sunset
-Fixed slider not moving while out of menu
-DayNightCycle files reorganised to fit into AOTTG2 project structure, namespaces added to scripts
-Unnecessary RPC calls removed, syncTimeRPC will be moved once OnSettingsCHanged event added
-time, daylength and paused values are loaded to TimeSettings.cs now

FAStre committed:
Sonarcloud fixes

FAStre commented:
Summary of changes:

1.  DayAndNightControl.cs : Script that controls the Time and Lighting(#113 ) system's mechanics, adapted and modified for use in AOTTG2 from a free unity store asset.
2.  TimeSwitcher.cs: Script attached to UI inputs that allows the user to edit the current time of  #113 
3.  DayLengthController.cs: Script attached to UI inputs that allows the user to edit the length of a cycle of #113 
4.  TimeSettings.cs: Script that stores the variables used for #113 , will be used along with RPC's once OnSettingsChanged event is added
5.  ToggleDayNightController.cs: Script that toggles #113 on and off.

Changes were made to all scenes, including changing lighting settings to allow for light to be influenced by colour settings rather than skybox. Additionally all ground/terrain/objects that represented a scene's 'floor' were given a standard name, "GroundTerrain" so that this component can always be grabbed by #113 and it's size used to ensure that #113 scales to fit any scene.

I do think that there are some changes that could be made in future to the time and lighting system, including features and enhancements, I'll make an issue for this soon.

FAStre committed:
Update DayAndNightControl.cs

re add rpc and small multiplayer compatibility fix

FransStreicher commented:
Summary of changes:

1.  DayAndNightControl.cs : Script that controls the Time and Lighting(#113 ) system's mechanics
2.  TimeSwitcher.cs: Script attached to UI inputs that allows the user to edit the current time of  #113 
3.  DayLengthController.cs: Script attached to UI inputs that allows the user to edit the length of a cycle of #113 
4.  TimeSettings.cs: Script that stores the variables used for #113 , will be used along with RPC's once OnSettingsChanged event is added
5.  ToggleDayNightController.cs: Script that toggles #113 on and off.

Changes were made to all scenes, including changing lighting settings to allow for light to be influenced by colour settings rather than skybox. Additionally all ground/terrain gameobjects in maps were given a standard name, "GroundTerrain" so that this component can always be grabbed by #113 and it's size used to ensure that #113 scales to fit any scene.

I do think that there are some changes that could be made in future to the time and lighting system, including features and enhancements, I'll make an issue for this soon.

FAStre committed:
Multiplayer compatibility fixes

FAStre committed:
comments added

FAStre committed:
Resolved merge conflict

FAStre committed:
Merge remote-tracking branch 'origin/development' into #113-Time-&-Lighting-System

FAStre commented:
Returning to issue now, tester feedback yielded the following recommendations:

- Make another sunrise colour, between the twilight and dawn, add a sunrise skybox. Can be done by adding another colour "sunrise" and another material for the sunrise skybox.

- Moon sometimes is appearing as if in front of terrain when it is behind/moon rendering issues.

- Adding a sunset skybox would maybe also be good, not absolutely needed, but if it looks good it's fine to add it

- Limit a lower limit of daylength, because it is actually potentially seizure inducing at low values, could be a risk to players with epilepsy? 

- Fix colossal map lighting 

- Implement SyncSettingsRpc to replace SyncTimeRPC 

The minor bugs like the system being turned off on game restart, and when restart happens that the daylength variable(controls a daynight cycles length) is reset are minor issues I'll just make a bug issue for, with very low priority.

FAStre committed:
upgraded to unity 2020.1f1 and compiler error fixes

FAStre committed:
Minor changes

-Minimum day length added
-Renamed some colour variables for more readability
-removed unneeded GameObject.Find

Lithrun reviewed (changes requested):
Request for review once you want me to review again

Commit:
Time of Day & Timecycle system.

FAStre committed:
Merge pull request #437 from WattanaGaming/development

Time of Day & Timecycle system.

FAStre commented:
A list of bugs/work still to be done:



- [x] 1. Change IN_GAME_MAIN_CAMERA (i think) to not use the skybox component, rather have everything set from rendersettings.skybox if possible, when that's done, changing the Day and Night Control script to change skybox rendersettings instead of the main camera scripts skybox component

- [x] 2.-Multiplayer compatibility @FransStreicher 

- [x] 3.-Investigate moving to a 24h range instead of from 0-1(the original asset has 0-1) this involves not only changing the Day and Night Control script but also many other user input controls, @WattanaGaming 

- [x] 4.- Moving to gradient lighting. Wattana you seem to have experience here, I can just point to which colours we already use that would be good to include @WattanaGaming 

- [x] 5.- Sorting out the moon issue bug, if you are keen to do this @WattanaGaming 

- [x] 6. -Allow DayLength Variable and currentTIme to be set from the prefab view, this will allow map makers to simply drag paste the prefab onto a scene and set their own default lighting settings that they prefer in the scene.

- Bug fixes that are high priority: 
- [x] 7.The Colossal map has strange floor lighting when the night time of the cycle comes, not sure what is the culprit.

- [x] 8. Utgard's floor seems a bit bright compared to other maps, not as much as colossal, not sure what is the culprit.

- [x] 9. Default Skybox bug fix, where the skybox appears to move strangely(time and lighting being activated does not cause this, culprit likely due to changes to IN_GAME_MAIN_CAMERA or camera changes I think.

- [x] 10. Non MC's should not be able to influence multiplayer settings at all, even for themselves. @FransStreicher 

- [x] 11.Colossal map city floor is now extremely bright during the day.

Commit:
Various fixes and updates

+ Switched from Skybox component to `RenderSettings.skybox`.

+ Fixed moon clipping issues

- Removed redundant Audio Listener from the moon camera

- Removed redundant lines that involve changing the skybox in `DayAndNightControl.cs`.

FIXME: Weird skybox movement. The problem disappears after toggling the day night cycle once.

Commit:
Updated the day night system to use timecycle

+ `currentTime` now uses 24-hour format,

+ `CurrentTime01` property is available for stuff that needs to old 0-1 format.

+ Also added a sky reflection probe to the main camera.

Also tidied up a small amount of codes.

To be continued in the next episode of Sleep Deprived Programming: obliterating `TimeOfDay.cs`(question marks?)

Author's comment: I dare you play the forest map and set the time to night. You'll feel the terror.

FAStre committed:
Added time settings and multiplayer setting synching

Additional changes include:
-Added a synctime button to synch settings, this is because I want to keep the time settings ui components as a unit, and not have to rely on any other synch buttons on the server settings page, because time settings I think will be moved to another place when the completed UI is finished.

-Fixed pause button

-Reduced moonlight intensity, however this requires some discussion before being final

atirutw commented:
> * Bug fixes that are high priority:
> * [ ]  6.The Colossal map has strange floor lighting when the night time of the cycle comes, not sure what is the culprit.
> * [ ]  7. Utgard's floor seems a bit bright compared to other maps, not as much as colossal, not sure what is the culprit.
> * [ ]  8. Default Skybox bug fix, where the skybox appears to move strangely(time and lighting being activated does not cause this, culprit likely due to changes to IN_GAME_MAIN_CAMERA or camera changes I think.
> * [ ]  9. Non MC's should not be able to influence multiplayer settings at all, even for themselves. @FransStreicher

I took a look at some of the lighting bugs described and here's the results of the initial investigation:

1. I could not confirm the strange lighting of the colossal map because I can't seem to find it in the map list when starting a game, however, the terrain floor is flickering. This is a bug with the lightmap for Unity's directional light when there's a large shadow-casting objects(for example: the terrain). Can be fixed by splitting up the terrain into smaller "tiles". Maybe we should just redo the terrain? The old one doesn't look very good anyway.
2. The direction light for the moon in Utgard is as bright as a sun(intensity is 1.4).

> * [ ]  4.- Moving to gradient lighting, adding sunrise and sunset orangish colours to. Wattana you seem to have experience here, I can just point to which colours we already use that would be good to include @WattanaGaming

You can consider this one done, btw.

Commit:
Mowed down some SonarCloud issues

Also made `TimecycleProfile` independent of `TimeOfDay`.

FAStre commented:
@WattanaGaming Thanks will look at those. Colossal map can be found from the Trost map, then select Colossal Titan as an option.

Commit:
Some updates

+ Fixed the funny lighting in the Colossal scene.

+ Lowered moon brightness in Utgard.

+ Updated `DayAndNightControl` in anticipation for an update to how the lighting is set up.

FAStre committed:
Day Night Cycle now permanently replaces all scene lighting

-Prefab added to all scenes, paused by default
-There is a pause/unpause toggle
-Added some slight adjustments to DayAndNightControl.cs that allows it to find some of the UI settings required for it to work
-Default daylength increased

FAStre committed:
Procedural skybox added

-Procedural skybox just uses the lighting values from the timecycle profile for now, but it does seem to look fine , just a slight red that lingers at the horizon even at midnight, should be changed?
-Moon now faces player camera again

Commit:
Stuff

+ Added option to not change the skybox

+ Linear color space.

Commit:
Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System

Commit:
Fixed weird skybox

Commit:
Revert to Forward rendering path

Commit:
Remvoed duplicate shader.

Commit:
Material update

Friendship ended with legacy shaders. Standard is now my friend.

Author's Comment: I don't event know how I forgot to extend the reflection probe for the sky. Fixed it.

Commit:
I don't know how I should name this anymore

So much stuff going on lmao.

FAStre committed:
Update New Timecycle Profile.asset

Commit:
Mostly `DayAndNightControl` updates

+ Fixed time progressing faster than it should

+ Exposed `currentTime` and `SecondsInAFullDay` in the editor

- Removed `DayLength` from `DayAndNightControl` due to simply being a redirect variable.

Commit:
Updated the skydome(moondome? lol) setup

Commit:
Slightly adjusted moon distance.

-10 on Z axis looks a bit too big so I changed it to -15

Commit:
Updated moon material

FAStre committed:
Added DayNightCycle to more scenes

-Utgard now starts at night
-All other maps start at day
-Some minor UI updates to make it time settings UI more readable and useable
-non mcs can no longer change time settings

Commit:
Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System

Commit:
Update DayAndNightControl.cs

+ `TimeOfDay()` renamed to `GetTimeOfDay()`, the name is now used by an enum. ``GetTimeOfDay()`` now also uses the said enum for easier code refactoring.

Commit:
Squashed some more SonarCloud code smells

I left the remaining parts out because they are multiplayer-related. especially the commented-out codes and unused photonView variable.

FAStre committed:
Updated UI backend

FAStre committed:
Update DayAndNightControl.cs

FAStre committed:
Added some UI validation

FAStre committed:
Small multiplayer improvement and removed unused materials

-MC's should now be able to synch settings as they are changed by inputs, rather than having to click a synch button

FAStre committed:
Sonarcloud fixes and code cleanup

-also reduced synch settings use and allowed slider movement in multiplayer while the daynight cycle is toggled to play

FAStre committed:
Merge branch 'development' into #113-Time-&-Lighting-System

FAStre committed:
Sonarcloud fixes

-also attempted to fix OnJoinedRoom not synching time\
-removed time syncbutton as it is no longer needed

FAStre committed:
Players joining a room now have their settings synched

-some sonarcloud fixes as well

FAStre committed:
removed default scene lighting and some nullref issues

FAStre committed:
Update DayAndNightControl.cs

FAStre commented:
@Lithrun Scripts/materials added for #113  can be found in Assets/Scripts/DayNightCycle and Assets/Materials/DayNightCycle. The DayNightCycle prefab is in the Resources folder. Changes to scenes are just lighting and other updates relevant to that.

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+            {
+                try
+                {
+                    MainCamera = GameObject.Find("MainCamera").GetComponent<Camera>();
```
Instead of allowing an exception, you can just check if it's null. Additionally, Camera.main should return the same object (https://docs.unity3d.com/ScriptReference/Camera-main.html)

FAStre replied:
Will change font and remove the unused prefab reference. @WattanaGaming would you be able to look at the leaves reflecting? If it is something to do with the leaf material/texture then I think it might be another issue not related to time and lighting, but if we can fix it with lighting settings then we should.

atirutw replied:
@FransStreicher fixed 👍

FAStre replied:
Strange, initially not catching a nullref caused issues for the non-mc, but now it doesn't anymore, will push the changes without the try catch here.

FAStre replied:
Ok checked on a rebuild and once again getting a nullref here, only for non-MC users though. For some reason the Main Camera is not loaded in time, but only on multiplayer and for non-MC's, so this is a workaround for that, and is necessary to avoid nullref errors.

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+                GameSettings.Time.currentTime = currentTime;
+                GameSettings.Time.dayLength = DayLength;
+                GameSettings.Time.pause = pause;
+                Debug.Log("Current Master Client time: " + GameSettings.Time.currentTime);
```
Avoid logging something every frame

FAStre replied:
It seems this is a leftover from debugging, will remove it.

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+        public Material skyBoxPROCEDURAL;
+
+        [Range(0f, 24f)] public float currentTime;
+        public float CurrentTime01 { get { return currentTime / 24; } set { currentTime = value * 24; } }
```
Always avoid using Numbers in your property / fields. We don't need a currentTime, currentTime1, currentTime2 ect.

FAStre replied:
Will do, @WattanaGaming would you like to tackle this as it is your implementation? I can also take it if you want, either way is fine with me.

atirutw replied:
`currentTime` is what we use for the 24 hour format and `CurrentTime01` is for 0-1 format as to not have `currentTime / 24` everywhere. Hopefully this is a good excuse lol.

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+
+
+            //Check if default light prefab exists, and if so, disable it
+            if (GameObject.Find("LightSet"))
```
So we need to find this gameObject twice?

FAStre replied:
We dont, will remove

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+        {
+            Service.Settings.SyncSettings();
+            //syncs settings after waiting a few seconds(
+            StartCoroutine(ExecuteAfterTime(5));
```
SyncSettings will synchronize **all** settings to everyone. Doing this when a player joins is one thing, why should do sync after 5 seconds again?

FAStre replied:
For some reason there seemed to be a bit of lag just as the player joined,(due to nullrefs  I think it threw off the synching of settings if the daynightcycle is running) so I added another one 5 seconds later and time seems to synch correctly. Will experiment again to see if it is no longer needed, as nullrefs were sorted out.

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+            }
+        }
+
+        public TimeOfDay GetTimeOfDay()
```
If TimeOfDay.UNKNOWN means that something went wrong, it will go wrong for when CurrentTime01:
- `0`
- `>= 0.1 && <= 0.2`
- `0.5`
- `0.75`
- `1`

Lithrun replied:
After checking in Visual Studio, this also appears to be unused?

FAStre replied:
@WattanaGaming  I think you added the unknown

atirutw replied:
@FransStreicher wdym? I only modified it to use enum

FAStre replied:
@WattanaGaming https://github.com/AoTTG-2/AoTTG-2/commit/97b2e408a1d91d1659ca7014c6ebff0192d2efcd here you changed it to enum and added the 'unknown' line of code, I'm wondering whether we need it? I can add a = sign to the > signs in the if statements so that time will always be one of the other values(morning,afternoon,evening,midnight), in that case, I don't know if we will need unknown as after adding equals signs in the if statements, any currentTime will always have a value 0<=x<=1 so perhaps no need for an unknown returning.

FransStreicher replied:
Seems I was overlooking that it's required, will just move the code of line to the top as Lithrun mentioned defaults are usually at the top

FAStre replied:
I think the line of code is needed and makes sense to be where it is, will just update some of the if conditions.

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+                if (frames == lightingUpdateInterval)
+                {
+                    frames = 0;
+                    UpdateLight();
```
If the game is paused, then why do we need to update the light every 10 frames?

FAStre replied:
Because the user can change the cycle even if the game is paused.

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+            return TimeOfDay.UNKNOWN; // If this return is reached, something probably went wrong
+        }
+        
+        public enum TimeOfDay
```
If there is a default / unknown value, it is usually placed at the top (0)

Lithrun commented on Assets/Photon Unity Networking/Resources/PhotonServerSettings.asset:
```diff
@@ -15,11 +15,11 @@ MonoBehaviour:
   PreferredRegion: 0
   EnabledRegions: -1
   Protocol: 0
-  ServerAddress: 51.210.5.100
```
Please don't change the server address and HostType

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+
+        [Range(0f, 24f)] public float currentTime;
+        public float CurrentTime01 { get { return currentTime / 24; } set { currentTime = value * 24; } }
+        public Slider TimeSlider  = null;
```
TimeSlider is unused

Lithrun replied:
TimeSlider is still unused

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+        public Slider TimeSlider  = null;
+        public Camera MoonCamera = null;
+        public Camera MainCamera = null;
+        public GameObject DayNightController = null;
```
DayNightController is unused

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+        [HideInInspector]
+        public float timeMultiplier = 1f; //how fast the day goes by regardless of the secondsInAFullDay var. lower values will make the days go by longer, while higher values make it go faster. This may be useful if you're siumulating seasons where daylight and night times are altered.
+        float lightIntensity; //static variable to see what the current light's insensity is in the inspector
+        public GameObject  SettingsUI = null;
```
timeMultiplier and SettingsUI are unused

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,257 @@
+            RenderSettings.skybox = skyBoxPROCEDURAL;
+            Service.Settings.OnTimeSettingsChanged += Settings_OnTimeSettingsChanged;
+            MoonCamera = GetComponentInChildren<Camera>();
+            //MainCamera =
```
MainCamera = ?

Lithrun commented on Assets/Scripts/DayNightCycle/DayLengthController.cs:
```diff
@@ -0,0 +1,58 @@
+        {
+            try 
+                {
+                    float dayLength =float.Parse(arg0);
```
Instead of exceptions, using float.TryParse here.

Lithrun commented on Assets/Scripts/DayNightCycle/DayLengthController.cs:
```diff
@@ -0,0 +1,58 @@
+
+        void Start()
+        {
+            ToggleDayNight = GameObject.Find("ToggleDayNightCycle").GetComponent<Toggle>();
```
ToggleDayNightCycle appears to be an UI element which is always present. You can just use a reference instead of finding it

Lithrun commented on Assets/Scripts/DayNightCycle/DayLengthController.cs:
```diff
@@ -0,0 +1,58 @@
+        void Start()
+        {
+            ToggleDayNight = GameObject.Find("ToggleDayNightCycle").GetComponent<Toggle>();
+            dayNightCycle = GameObject.Find("Day and Night Controller").GetComponent<DayAndNightControl>();
```
So does this mean that every scene requires this prefab? If so, we should document it somewhere..

FAStre replied:
Yes, I left a comment on #365 that once #113 is done, that it should be included on every scene, and added to the mapping prefabs folder for mappers.

Lithrun commented on Assets/Scripts/DayNightCycle/SyncTimeButtonHandler.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace Assets.Scripts.DayNightCycle
+{
+    class SyncTimeButtonHandler : MonoBehaviour
```
Is this class even used? I can't seem to find any references in the scenes I tested it with. (City, Outside the Walls)

FAStre replied:
It used to be the syncbutton script, will remove it.

Lithrun commented on Assets/Scripts/DayNightCycle/TimeSwitcher.cs:
```diff
@@ -0,0 +1,84 @@
+       void Start()
+       {
+            
+            ToggleDayNight = GameObject.Find("ToggleDayNightCycle").GetComponent<Toggle>();
```
Once again, this ToggleDayNightCycle appears to always be present. Why use .Find instead of a reference via the inspector?

FAStre replied:
Will do

Lithrun commented on Assets/Scripts/DayNightCycle/TimeSwitcher.cs:
```diff
@@ -0,0 +1,84 @@
+                {
+                 try 
+                    {
+                    double seconds = System.TimeSpan.Parse(time).TotalSeconds;
```
Once again, use .TryParse so no need to throw an exception here

Lithrun commented on Assets/Scripts/DayNightCycle/TimeSwitcher.cs:
```diff
@@ -0,0 +1,84 @@
+
+        public void OnDrag(PointerEventData eventData)
+        {
+            if (PhotonNetwork.isMasterClient)
```
This is a very expensive way of synchronizing time. Whenever this event is triggered, you are sending **all** settings to everyone in the lobby. Just by slowly dragging the bar I've sent +120 RPCs to the photon server. Settings should only be synchronized whenever the "Sync" button is pressed.

Each SyncSettings RPC is 5KB in data alone, so within just a second I've sent 600KB of data to the Photon Server, which it then also needs to send to every player.

FAStre replied:
It is a tradeoff then from UX for less network use, I can put it back to just using a synch button that has to be tapped every time the slider is dragged by the user. The reason I changed it to onDrag was for UX, and I thought that once SyncSettings is expanded to only only send specific settings, so for example just time settings, that the network use would be significantly reduced. 

Alternatively we could use OnEndDrag? That way it only synchs when the slider stops being dragged, which would be equal to what would be sent once the synch button was tapped anyway.

Lithrun commented on Assets/Scripts/DayNightCycle/ToggleDayNightController.cs:
```diff
@@ -0,0 +1,60 @@
+        public Color DefaultSkyColor;
+        public Color DefaultEquatorColor;
+        public Color DefaultHorizonColor;
+        public GameObject DayNightControllerPrefab = null;
```
DayNightControllerPrefab, PauseDayNightButton, DayNightController, MainLight, skyBoxReset, time and seconds are all unused.

FAStre replied:
Yes will remove these.

Lithrun commented on Assets/Scripts/DayNightCycle/ToggleDayNightController.cs:
```diff
@@ -0,0 +1,60 @@
+
+        void ToggleValueChanged(Toggle change)
+        {
+            if (PhotonNetwork.isMasterClient)
```
We should only sync settings once we click the synchronize button

FAStre replied:
This is again a UX decision, the user would have to anyway tap a synch button for this rpc to take effect, and this one only synchs once a toggle, so it would be no more RPC's being sent than it would with a synch button being pressed after changing the toggle value.

FAStre replied:
Well even ignoring UX, sending it on ToggleValueChanged is the same as not having it and tapping the synch button(which some users may tap more than once) after changing the toggle setting. By having it here it makes it easier to use and sends the same/less  RPC data anyway.

Lithrun commented on Assets/Scripts/Time of Day/TimecycleProfile.cs:
```diff
@@ -0,0 +1,38 @@
+namespace Assets.Scripts.DayNightCycle
+{
+    [CreateAssetMenu(fileName = "New Timecycle Profile", menuName = "Timecycle Profile")]
+    class TimecycleProfile : ScriptableObject
```
Access modifier is missing

FAStre replied:
@WattanaGaming

atirutw replied:
Whoops. Will fix it right away

Lithrun reviewed (changes requested):
Overall feedback:
- Name properties & fields correctly, using PascalCase. Your visual studio should give feedback about this based on our .editorconfig file. Don't use CAPSLOCK EITHER for anything
- Remove the lfs/objects files. Don't think we need those

For code feedback, see comments 
Based on testing:
Day Night Cycle > ToggleDayNightCycle gameObject has missing reference
![image](https://user-images.githubusercontent.com/22615133/104755740-22a7e100-575b-11eb-97ba-189775900e9c.png)
The UI text is black. This makes it very difficult to see what is written on the UI 
![image](https://user-images.githubusercontent.com/22615133/104755877-508d2580-575b-11eb-908c-60d3271ee574.png)
Leaves are reflecting light
![image](https://user-images.githubusercontent.com/22615133/104756227-d3ae7b80-575b-11eb-90c6-dacedb56e355.png)

Commit:
Fixed glossy tree leaves

Commit:
Added access modifier to the `TimecycleProfile` class

Lithrun committed:
Added ModifiedDate to the TimeSettings

Lithrun committed:
Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System

FAStre committed:
Code review changes

-Removed lfs files
-Removed unused references
-Reduced network usage by using OnEndDrag to synch settings with the TimeSlider
-Used Tryparse isntead of try catch
-Added 60 second minimum daylength to DayAndNightControl.cs too to prevent an abusive MC from overriding client minimum daylength
-Some code optimization too and variable name changing

Commit:
Added stardome(not working)

Idk how to make this work with PUN.

@FransStreicher please look into this when you're awake.

FAStre committed:
Update Day and Night Controller.prefab

-added stardome to moon layer

Toorah committed:
Added Stars to Day Night Cycle

- Added custom sky shader based on built-in sky
- Added 4D Voronoi Noise CGInc
- Refactored Property Naming
- Added Sky Material Controls in DayAndNightControl.cs

FAStre committed:
Resolve Merge Conflict

FAStre commented:
Ok it seems like there's a bit more to do.

- [x] Moving multiplayer compatibility to use OnJoinedRoom has made multiplayer settings not synch anymore, whereas before it was working fine. Using OnJoinedRoom is however a cleaner solution so it should be investigated why this isnt working with time settings @FransStreicher 


- [x] Re-add the SynchTime button, instead of people using enter on the input fields, rather keep it as a button to submit all time input fields. @FransStreicher -After consulting with testers, for now enter for submit will stay, when UI is reworked project wide it may change to a button or not.

- [x] Fixing the shadow flickering on some maps. @WattanaGaming you mentioned looking into this. -will be solved by #526

Toorah commented:
> *  Integrate sky, moon and stars rendering to just one camera, the main camera(and remove MoonCamera), should improve performance @Toorah you seemed to be keen to work on this.

Yes, I can do that maybe tomorrow

> *  Implementing basic clouds (optional) @Toorah you mentioned working on this.

I tested something yes, though I think it will need more time, and I don't know if it's rather linked to the Weather System

FAStre commented:
> > * Integrate sky, moon and stars rendering to just one camera, the main camera(and remove MoonCamera), should improve performance @Toorah you seemed to be keen to work on this.
> 
> Yes, I can do that maybe tomorrow
> 
> > * Implementing basic clouds (optional) @Toorah you mentioned working on this.
> 
> I tested something yes, though I think it will need more time, and I don't know if it's rather linked to the Weather System

Cool, if clouds are better kept in #112 , then I dont think we have to add it now, the stars that have been recently added already improve the sky enough for a preview release/devblog.

Commit:
Fixed moon camera desync with Rotation Constraint

No more motion-sickness inducing sky!

Commit:
Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System

Commit:
Fixed weird night lighting in Outside map and slightly adjusted moon size

Commit:
Updated Day and Night controller related stuff

+ Reduced Lighting Update Interval to 5 frames.

+ Lighting Update Interval now also applies to when the day/night cycle is progressive instead of only when paused

+ Replaced `MainCamera` and `GameObject.Find("MainCamera").GetComponent<Camera>();` in `DayAndNightControl` with `Camera.main`

+ Rewrote some confusing comments

- Remove commented codes

+ Improved newline consistency

atirutw commented:
> * [ ]  Integrate sky, moon and stars rendering to just one camera, the main camera(and remove MoonCamera), should improve performance @Toorah you seemed to be keen to work on this.(optional)

This is impossible without the moon getting obscured by fog.

Moon behind everything = gets obscured by fog,
Moon close up and small = renders in front of everything

The only way (that I know) to make the moon render behind everything without getting obscured by the fogs is to use a second camera.

Commit:
Code smells

Commit:
Update DayAndNightControl.cs

FAStre committed:
Update DayAndNightControl.cs

-for multiplayer testing

FAStre committed:
Networking changes + UI bug fixes

-Non-MC now gets correct time settings onJoinedRoom
-Time settings UI is cleared when leaving a scene, no leftover values in input fields etc
-SyncTime button removed
-Lighting Update interval reduced to 1 for more smoothness, doesnt impact fps but increases cosmetics alot

FAStre committed:
Merge branch 'development' into #113-Time-&-Lighting-System

atirutw commented:
Some quality of life stuff I wanna implement. They're optional but will probably help feed my ego

- [x] Make `DayAndNightControl.cs` preview changes in editor using `OnValidate()`.
- [x] Try to make the codes look even better and more readable.

Commit:
Fixed forest ground texture

Commit:
Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System

FAStre committed:
Update .gitignore

-csproj files are auto generated, no need to track as far as  i know

Commit:
Added .csproj extension to .gitignore

Commit:
Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System

Commit:
Well that was a clusterfuck isn't it?

Commit:
Revert "Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System"

This reverts commit 2c8db1c20eeb6e30fdc79ac7e0414952003522a7, reversing
changes made to c000d713ecd12ee7c8a21c70a43295b58c2100ca.

Commit:
Added .csproj files

WattanaGaming commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,247 @@
+        public float DayLength = 300f; //default value is 300 seconds in one day
+        public bool pause { get; set; }
+        
+        [HideInInspector]
+        float lightIntensity; //static variable to see what the current light's insensity is in the inspector
```
> "to see what the current light's insensity is in the inspector"

> Uses a private float that have the `HideInInspector` attribute

This feel so wrong on many levels

atirutw replied:
@FransStreicher are you sure you're doing this one right? You want it to be visible in the editor but you used a private float and gave it the `HideInInspector` attribute on top of that.

Commit:
Make `DayAndNightControl.cs` preview changes in editor using `OnValidate()`

Plus some improvements to readability

Commit:
Update DayAndNightControl.cs

Commit:
Make `DayNightControl` wait for the main camera before adding constraint

FAStre committed:
Sonarcloud fixes

FAStre committed:
Resolved some conflicts in AoTTG2.unity

FAStre committed:
Update DayAndNightControl.cs

-Fixed non-MC time synching while MC has daynight cycle unpaused

FAStre committed:
added some time input validation

FAStre committed:
Update DayAndNightControl.cs

-small sonarcloud fix

FAStre committed:
Update PhotonServerSettings.asset

-readded removed address

FAStre committed:
Update PhotonServerSettings.asset

-removed no longer existing RPCs

Commit:
Upgraded lighting preview for `DayAndNightControl`

Lighting preview now properly sets the lighting settings in `RenderSettings` and updates the reflection using a reflection probe that follows the scene view camera.

Commit:
Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System

Commit:
Exclude codes involving `UnityEditor` namespace from build

FAStre committed:
Update DayAndNightControl.cs

-This line doesnt have to be in Update, only relevant to when settings change

Toorah committed:
Created real prefabs for "The City I" for buildings, and enabled static batching

Toorah committed:
Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System

Commit:
Update The City I.unity

Commit:
Dirty fix to keep CI build from failing

Why tf is our CI including editor scripts anyway?

Commit:
Update PrefabReplacer.cs

Toorah committed:
Update PrefabReplacer.cs

- Make it workjwith < 2020.2

Toorah committed:
Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System

Commit:
Update OutSide.unity

Commit:
Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System

Lithrun commented on Assets/Editor/PrefabReplacer.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace Assets.Editor.Prefabs
+{
+    public class PrefabReplacer : EditorWindow
```
PrefabReplacer seems to have the same functionality as the "GameObjectToPrefabConverter", please merge the two together instead of having two (near) identical tools

atirutw replied:
@Toorah you made it so.... *shrugs*

Lithrun commented on Assets/Photon Unity Networking/Resources/PhotonServerSettings.asset:
```diff
@@ -15,7 +15,7 @@ MonoBehaviour:
   VoiceAppID: 
   ChatAppID: 
-  HostType: 2
+  HostType: 0
```
Please don't change the host type

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+        [SerializeField] private float sunRotationOffset = 0f;
+        [Tooltip("The amount of frames to wait before doing the next lighting update")]
+        [SerializeField] private int lightingUpdateInterval = 10;
+        public Material skyBoxPROCEDURAL;
```
Apply PascalCase

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+
+namespace Assets.Scripts.DayNightCycle
+{
+    public class DayAndNightControl : MonoBehaviour
```
Public properties & fields are always PascalCase

Lithrun replied:
Also remove the default unity comments

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+        public bool pause { get; set; }
+        public float lightIntensity; //static variable to see what the current light's insensity is in the inspector
+
+        Camera targetCam;
```
targetCamera appears to be unused, also missing its access modifier

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+        private int frames;
+       
+        // Use this for initialization
+        void Start()
```
Always be explicit when it comes to access modifiers

FAStre replied:
Does even the internal unity monobehaviours of a class need access modifiers? The class itself has one so idk if void Start() and void Update need one too, havnt seen any other ones in the project with access modifiers.

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+            UpdateLightingSettings();
+            UpdateLight(); // Initial lighting update.
+
+            IEnumerator SetupCameraConstraintAndReflectionProbe()
```
Local Functions are always at the bottom of the method

FAStre replied:
@WattanaGaming  I think you added this would you like to take it up?

atirutw replied:
That was my habit of putting local functions right above where they're used. Will fix

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+        public Camera MainCamera = null;
+        public int currentDay = 0;
+        public Light directionalLight;
+        public float DayLength = 300f; //default value is 300 seconds in one day
```
DayLength = 300f. Default value is 300, I can see that. Rather add a /// summary instead of a comment, or remove it

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+                //time according to a 24h time range
+                if (currentTime > 24)
+                {
+                    currentTime = currentTime - (24 * (int)Math.Floor(currentTime/24));
```
Could use a `currentTime -=` instead of `currentTime = currentTime -`

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+            if (!pause)
+            {
+                currentTime += (Time.deltaTime / DayLength) * 24;
+                if (CurrentTime01 >= 1) // If CurrentTime >= 24 hours
```
CurrentTime01 is still a vague property

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+            RenderSettings.sun = directionalLight; // Procedural skybox needs this to work
+            RenderSettings.fog = true;
+
+            if (timecycle)
```
if statement can be inverted to make the code more readable => `if (!timeCycle) return;`

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+            directionalLight.transform.Rotate(Vector3.up, sunRotationOffset - 90, Space.World);
+            moon.transform.forward = -directionalLight.transform.forward;
+
+            if (timecycle)
```
#InvertThatIfStatement

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+        [Range(0f, 24f)] public float currentTime;
+        public float CurrentTime01 => currentTime / 24;
+        public Slider TimeSlider  = null;
+        public Camera MoonCamera = null;
```
Shouldn't the MoonCamera be disabled if you are within a time of day where the moon is not present? In your Update method, you are changing the FOV of the moon every frame, yet what is the point if the moon is not visible?

FAStre replied:
I can put an if statement for the times when the moon is definitely below terrain and unobservable, then the FOV is only updated when the moon is visible

atirutw replied:
> I can put an if statement for the times when the moon is definitely below terrain and unobservable, then the FOV is only updated when the moon is visible

Does this not also effect the sun?

FAStre replied:
It does seem so from some quick testing, so i dont think we can disable this.

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,246 @@
+            }
+        }
+
+        void OnValidate()
```
"This function is called when the script is loaded or a value is changed in the Inspector (Called in the editor only)."

So, since this can only be called by the editor, where is there an extra #if UNITY_EDITOR region?

FAStre replied:
@WattanaGaming Your addition as well, please check it out if you have time to.

Lithrun commented on Assets/Scripts/DayNightCycle/DayLengthController.cs:
```diff
@@ -0,0 +1,56 @@
+
+        void Start()
+        {
+            dayNightCycle = GameObject.Find("Day and Night Controller").GetComponent<DayAndNightControl>();
```
Let's not use hardcoded strings to find something. Where are the references?

FAStre replied:
I don't know if I can reference the Day and Night Controller because it is added to every scene, so basically the only way to find it is when the UI component is enabled in a scene, to then find the local Day and Night Controller game object.

Also to note DayAndNightControl.cs itself is about 0.3% of CPU ms usage so doesn't significantly affect performance. Lighting and Rendering is by far the larger culprit of performance drops thus far I think(from unity profiler)

Lithrun commented on Assets/Scripts/DayNightCycle/DayLengthController.cs:
```diff
@@ -0,0 +1,56 @@
+using UnityEngine.SceneManagement;
+namespace Assets.Scripts.DayNightCycle
+{
+    public class DayLengthController : MonoBehaviour
```
Access modifiers are great

Lithrun commented on Assets/Scripts/DayNightCycle/DayLengthController.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            if (PhotonNetwork.isMasterClient)
+            {
+                var se = new InputField.SubmitEvent();
```
We don't need to do this every frame. Could just be moved to Start / Awake

Lithrun commented on Assets/Scripts/DayNightCycle/DayLengthController.cs:
```diff
@@ -0,0 +1,56 @@
+        //grabbing the local scene's DayAndNightControl script, and adding a scene changed listener
+        void OnEnable()
+        {
+            dayNightCycle = GameObject.Find("Day and Night Controller").GetComponent<DayAndNightControl>();
```
References instead of .Find

FAStre replied:
I don't know if I can reference the Day and Night Controller because it is added to every scene, so basically the only way to find it is when the UI component is enabled in a scene, to then find the local Day and Night Controller gameobject.

Also to note DayAndNightControl.cs itself is about 0.3% of CPU ms usage so doesn't significantly affect performance. Lighting and Rendering is by far the larger culprit of performance drops thus far I think(from unity profiler)

Lithrun commented on Assets/Scripts/DayNightCycle/DayLengthController.cs:
```diff
@@ -0,0 +1,56 @@
+        void OnEnable()
+        {
+            dayNightCycle = GameObject.Find("Day and Night Controller").GetComponent<DayAndNightControl>();
+            SceneManager.sceneLoaded += OnSceneLoaded;
```
If we subscribe to an event OnEnable, then we also need to UnSubscribe OnDisable

Lithrun commented on Assets/Scripts/DayNightCycle/TimeSwitcher.cs:
```diff
@@ -0,0 +1,86 @@
+        {
+            dayNightCycle = GameObject.Find("Day and Night Controller").GetComponent<DayAndNightControl>();
+            TimeSlider.value = dayNightCycle.CurrentTime01;
+            SceneManager.sceneLoaded += OnSceneLoaded;
```
Don't forget to UnSubscribe, otherwise we will keep subscribing whenever this is enabled

Lithrun commented on Assets/Scripts/DayNightCycle/ToggleDayNightController.cs:
```diff
@@ -0,0 +1,51 @@
+        void OnEnable()
+        {
+            dayNightCycle = GameObject.Find("Day and Night Controller").GetComponent<DayAndNightControl>();
+            SceneManager.sceneLoaded += OnSceneLoaded;
```
Don't forget to UnSubscribe and click that dislike button. Also checkout my Patreon to get money

Lithrun commented on Assets/Scripts/Settings/TimeSettings.cs:
```diff
@@ -0,0 +1,34 @@
+                case Difficulty.Hard:
+                case Difficulty.Abnormal:
+                case Difficulty.Realism:
+                    dayLength = 1000;
```
All settings should be assigned a value, otherwise you will get NullReferenceExceptions

Lithrun commented on Assets/Scripts/Settings/TimeSettings.cs:
```diff
@@ -0,0 +1,34 @@
+        public float? currentTime { get; set; }
+        public float? dayLength { get; set; }
+        public bool? pause { get; set; }
+        public DateTime LastModified { get; set; }
```
Must be of type DateTime?

Lithrun commented on Assets/Scripts/UI/InGame/ServerSettingsPage.cs:
```diff
@@ -1,4 +1,5 @@
+using Assets.Scripts.Settings.Gamemodes;
 using System.Collections.Generic;
 using System.Linq;
 using UnityEngine;
```
Lots of random newlines, seems unnecessary

Lithrun commented on ProjectSettings/ProjectVersion.txt:
```diff
@@ -1,2 +1,2 @@
-m_EditorVersion: 2020.1.17f1
-m_EditorVersionWithRevision: 2020.1.17f1 (9957aee8edc2)
+m_EditorVersion: 2020.2.1f1
```
Should keep in mind that 2020.2.1f1 is not possible until #373

Lithrun reviewed (commented):
Alright Pull Request review time... 166 files, fun time~

For individual code comments, see below. Overall:

**Quality Settings - Shadows**
With the introduction of this issue, the nights will look a lot darker due to the shadows, which is great. Yet, is it intentional that shadows can be disabled on the lowest quality settings? If so, playing on a night mode with disabling shadows gives them a significant advantage:

Low Settings
![image](https://user-images.githubusercontent.com/22615133/107280436-a69e7180-6a58-11eb-8993-7e2c2b2b308b.png)

Ultra settings
![image](https://user-images.githubusercontent.com/22615133/107280545-c5046d00-6a58-11eb-9f5d-560437673241.png)

**Performance**
Haven't been able to test in depth with profiler, but will see what testing team reports first.

**Other**
- Assets/New Terrain 1.asset what does this terrain asset do? Leftover? If so, delete
- Assure that there are always access modifiers
- What are these files inside "hooks/" ?

Commit:
PascalCase for skybox

Commit:
Removed unused targetCam variable

Commit:
Unsubscribed and disliked

Commit:
Moved SetupCameraConstraintAndReflectionProbe to bottom

Commit:
Use a currentTime -= instead of currentTime = currentTime -

Commit:
Quarantine OnValidate to editor only

Commit:
More unsubscribes

Commit:
PascalCase

Commit:
Update Day and Night Controller.prefab

Commit:
even more PascalCase

FAStre committed:
Moved unsubscribe code

- Putting it in OnDisable() results in the OnSceneLoaded() method not being called, which is required to reset the time settings UI so that old settings that arnt in effect dont remain in the input boxes and toggle.

FAStre committed:
Merge branch 'development' into #113-Time-&-Lighting-System

FAStre committed:
Resolved build failing error and set utgard default back to night

FAStre committed:
Update PhotonServerSettings.asset

FAStre committed:
Update DayAndNightControl.cs

Commit:
URP Upgrade

Commit:
Update ParticleScaling.cs

Commit:
Update DayAndNightControl.cs

Commit:
Enabled AA on main camera

Commit:
Purple terrain fix

FAStre committed:
Utgard time reset to night+some minor graphics settings adjustments

Commit:
Toorah stars shader knockoff + night sky rendering fix

Commit:
Merge branch '#113-Time-&-Lighting-System' of https://github.com/AoTTG-2/AoTTG-2 into #113-Time-&-Lighting-System

Commit:
Shiny hairs fix

FAStre committed:
Merge branch 'development' into #113-Time-&-Lighting-System

Commit:
Double sided shader for house fight

FAStre committed:
Removed default unity comments

FAStre committed:
Readability improvements+code review changes

FAStre committed:
Revert changes to csproj files from URP upgrade

FAStre committed:
Setting branch back to pre-URP upgrade

FAStre committed:
Update ProjectVersion.txt

-Downgrade to 202.1.17f1

FAStre committed:
Readability and code review changes

FAStre committed:
code smells

Commit:
Scene view nullref fix

FAStre committed:
returning to URP upgrade + unity 2020.2

Reverting branch to commit e806c546

FAStre committed:
Upgrade to 2020.2.5f1

FAStre committed:
Merge branch 'development' into #113-Time-&-Lighting-System

FAStre committed:
Packages+manifest merge changes

Commit:
Missing namespace fix

AoTTG2.core,asmdef is imposter very sus

Commit:
Code smell

Commit:
Improved night sky

- Removed duplicate shaders folder
- Stars now move like they should IRL. :open_mouth:
- My attempt at randomizing the brightness for stars looked ugly so I removed it entirely.

Toorah committed:
Updated Starry Sky Shader for URP

- the CGINC path needed to be relative (vs absolute within the project)

FAStre committed:
Added temporary HideHUD toggle for filming purposes

-To be reverted after filming preview footage
-F1 used to toggle HUD on/off

FAStre committed:
Merge branch 'development' into #113-Time-&-Lighting-System

FAStre committed:
Update ServerSettingsPage.cs

Commit:
Update Day and Night Controller.prefab

FAStre committed:
Update IN_GAME_MAIN_CAMERA.cs

-to be reverted after footage taken for preview

Commit:
Chinese leaves are now double sided

FAStre committed:
Slight multiplayer fix

-gamesettings need a default value, so if the MC doesnt synch settings(via changing them from the time settings input) before a non-MC player joins, then a error will come up preventing player spawn from the unity editor

FAStre committed:
Update PhotonServerSettings.asset

Lithrun committed:
Resolve multiplayer issue

FAStre committed:
Merge branch 'development' into #113-Time-&-Lighting-System

FAStre committed:
Merge branch 'development' into #113-Time-&-Lighting-System

FAStre committed:
Update track - akina.unity

FAStre committed:
Readd DayNightController to Akina

FAStre committed:
Update track - akina.unity

Lithrun committed:
Fix barrier not being destroyed on late join
- But, it does disappear before the countdown on a restart for the non-MC

FAStre committed:
resolve merge conflicts

FAStre committed:
Some more merge resolving

Commit:
Partially fix broken character

TODO: missing facial texture

FAStre committed:
Fix Levi eyes and face material

FAStre committed:
Fix akina barriers

FAStre committed:
Custom difficulty working on singleplayer

-current testing build has this already

FAStre committed:
Code smells

-removed unused using

FAStre committed:
Fixed titan health label

FAStre committed:
Fixed DayLength not synching right and nonMC issue

FAStre committed:
Variable Renaming

FAStre committed:
Fix NonMC bug

FAStre committed:
Remove city occlusion culling

Lithrun commented on Assets/Scripts/Gamemode/GamemodeBase.cs:
```diff
@@ -172,6 +172,9 @@ public virtual TitanConfiguration GetTitanConfiguration(MindlessTitanType type)
 
         public virtual void OnRestart()
         {
+            Debug.Log("CLEARING THAT CASH");
```
Remove old comment

Lithrun commented on Assets/Scripts/Gamemode/RacingGamemode.cs:
```diff
@@ -99,9 +102,20 @@ private void RequestStatus(PhotonMessageInfo info)
         private void RacingStartRpc(PhotonMessageInfo info)
         {
             if (!info.sender.IsMasterClient) return;
+            Debug.Log(nameof(RacingStartRpc));
```
Don't need log comments like those

Lithrun commented on Assets/Scripts/Gamemode/RacingGamemode.cs:
```diff
@@ -99,9 +102,20 @@ private void RequestStatus(PhotonMessageInfo info)
-            UiService.ResetMessage(LabelPosition.Center);
+            FindObjectsOfType<RacingStartBarrier>().ToList()
+                .ForEach(x => x.gameObject.SetActive(false));
+            Debug.Log("Running Couroutine");
```
nor this

Lithrun commented on Assets/Scripts/Settings/TimeSettings.cs:
```diff
@@ -0,0 +1,34 @@
+       
+        public float? currentTime { get; set; }
+        public float? dayLength { get; set; }
+        public bool? pause { get; set; }
```
Current time and pause do not have a default setting, thus there may be a null reference exception.

FAStre replied:
Currently set from the prefabs variables, of which maybe map settings are better, will check that.

Lithrun commented on Assets/Scripts/Settings/TimeSettings.cs:
```diff
@@ -0,0 +1,34 @@
+    public class TimeSettings
+    {
+       
+        public float? currentTime { get; set; }
```
public keyword is always **P**ascal**C**ase. This isn't Java >.>

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,248 @@
+            //MC loads settings on start because these values need to have a default.
+            if (PhotonNetwork.isMasterClient)
+            {
+                GameSettings.Time.currentTime = CurrentTime;
```
These are either supplied by the Difficult or the users custom settings. Because of the current implementation: the day night cycle cannot be paused, start with a specific time or specify what the length of a day should be by a global, gamemode or map setting. Remove such limitation to the fancy system.

FAStre replied:
Sounds good, at the moment it is sort of relegated to map setting, except via the prefab, so I can see if there is a place to set map settings and make it so the prefab allows the editing of those settings to for ease of use for mappers.

Lithrun replied:
Should still be via the TimeSetting, as otherwise it won't work well for custom maps, nor for my "Lithruns Hardcore Capture Gamemode Night 24/7 Edition" difficulty setting.

FAStre replied:
I see yeah, we can discuss the implementation over VC sometime.

Lithrun commented on Assets/Scripts/DayNightCycle/DayAndNightControl.cs:
```diff
@@ -0,0 +1,248 @@
+                ConstraintSource source = new ConstraintSource();
+                Debug.Log("Waiting for Main Camera...");
+                yield return new WaitUntil(() => MainCamera != null);
+                Debug.Log("The main camera has arriveth");
```
These logs don't appear that useful

Lithrun reviewed (changes requested):
Final comments. Good luck

FAStre committed:
Code review chanes

FAStre committed:
Fix coroutine

FAStre committed:
Merge enclosing if statements

FAStre committed:
Change Time Settings setup

FAStre committed:
Update TimeSettings.cs

FAStre committed:
Remove commented code

Lithrun reviewed (approved):
Alright, time to merge this one 😄
